File: turms-client-swift/Sources/TurmsClient/Driver/Service/HeartbeatService.swift
import Foundation

class HeartbeatService: BaseService {
    private static let HEARTBEAT_FAILURE_REQUEST_ID: Int64 = -100
    private static let HEARTBEAT_REQUEST = Data([0])

    private let lock = Lock()

    private let heartbeatInterval: TimeInterval
    private let heartbeatTimerInterval: TimeInterval
    private var lastHeartbeatRequestDate: TimeInterval = 0
    private var heartbeatTimer: Timer?
    private var heartbeatContinuations: [UnsafeContinuation<Void, any Error>] = []

    init(stateStore: StateStore, heartbeatInterval: TimeInterval? = nil) {
        self.heartbeatInterval = heartbeatInterval ?? 120
        heartbeatTimerInterval = max(1, self.heartbeatInterval / 10)
        super.init(stateStore)
    }

    var isRunning: Bool {
        return heartbeatTimer?.isValid == true
    }

    func start() {
        lock.locked {
            if isRunning {
                return
            }
            heartbeatTimer = Timer.scheduledTimer(
                timeInterval: heartbeatInterval,
                target: self,
                selector: #selector(checkAndSendHeartbeat),
                userInfo: nil,
                repeats: true
            )
        }
    }

    func stop() {
        heartbeatTimer?.invalidate()
    }

    func send() async throws {
        return try await withUnsafeThrowingContinuation { continuation in
            if !stateStore.isConnected || !stateStore.isSessionOpen {
                return continuation.resume(throwing: ResponseError(code: .clientSessionHasBeenClosed))
            }
            Task {
                guard let tcp = stateStore.tcp else {
                    return continuation.resume(throwing: ResponseError(code: .clientSessionHasBeenClosed))
                }
                do {
                    try await tcp.write(HeartbeatService.HEARTBEAT_REQUEST)
                } catch {
                    return continuation.resume(throwing: error)
                }
                lock.locked {
                    self.heartbeatContinuations.append(continuation)
                }
            }
        }
    }

    func fulfillHeartbeatContinuations() {
        lock.locked {
            repeat {
                heartbeatContinuations.popLast()?.resume()
            } while !heartbeatContinuations.isEmpty
        }
    }

    func rejectHeartbeatContinuationsIfFail(_ notification: TurmsNotification) -> Bool {
        if notification.hasRequestID, notification.requestID == HeartbeatService.HEARTBEAT_FAILURE_REQUEST_ID {
            rejectHeartbeatContinuations(ResponseError(notification))
            return true
        }
        return false
    }

    func rejectHeartbeatContinuations(_ error: Error) {
        lock.locked {
            repeat {
                heartbeatContinuations.popLast()?.resume(throwing: error)
            } while !heartbeatContinuations.isEmpty
        }
    }

    @objc func checkAndSendHeartbeat() async throws {
        let now = Date().timeIntervalSince1970
        let difference = min(now - stateStore.lastRequestDate.timeIntervalSince1970,
                             now - lastHeartbeatRequestDate)
        if difference > heartbeatInterval {
            do {
                try await send()
            } catch {
                // TODO: log
                return
            }
            lastHeartbeatRequestDate = max(lastHeartbeatRequestDate, now)
        }
    }

    override func close() async {
        onDisconnected()
    }

    override func onDisconnected(_ error: Error? = nil) {
        stop()
        rejectHeartbeatContinuations(ResponseError(code: .clientSessionHasBeenClosed, cause: error))
    }
}

File: turms-client-dart/lib/src/driver/service/heartbeat_service.dart
import 'dart:async';
import 'dart:math';

import '../../exception/response_exception.dart';
import '../../model/proto/notification/turms_notification.pb.dart';
import '../../model/response_status_code.dart';
import '../state_store.dart';
import 'base_service.dart';

class HeartbeatService extends BaseService {
  static const _defaultHeartbeatIntervalMillis = 120 * 1000;
  static const _heartbeatFailureRequestId = -100;
  static const List<int> _heartbeatRequest = [0];

  final int _heartbeatIntervalMillis;
  final Duration _heartbeatTimerInterval;
  int _lastHeartbeatRequestDate = 0;
  Timer? _heartbeatTimer;
  final List<Completer<void>> _heartbeatCompleters = [];

  HeartbeatService(StateStore stateStore, int? heartbeatIntervalMillis)
      : _heartbeatIntervalMillis =
            heartbeatIntervalMillis ?? _defaultHeartbeatIntervalMillis,
        _heartbeatTimerInterval = Duration(
            milliseconds: max(
                1,
                (heartbeatIntervalMillis ?? _defaultHeartbeatIntervalMillis) ~/
                    10)),
        super(stateStore);

  bool get isRunning => _heartbeatTimer?.isActive == true;

  void start() {
    if (isRunning) {
      return;
    }
    _heartbeatTimer = Timer.periodic(
      _heartbeatTimerInterval,
      (_) {
        final now = DateTime.now().millisecondsSinceEpoch;
        final difference = min(
            now - stateStore.lastRequestDate, now - _lastHeartbeatRequestDate);
        if (difference > _heartbeatIntervalMillis) {
          send();
          _lastHeartbeatRequestDate = now;
        }
      },
    );
  }

  void stop() => _heartbeatTimer?.cancel();

  Future<void> send() async {
    if (!stateStore.isConnected || !stateStore.isSessionOpen) {
      throw ResponseException(
          code: ResponseStatusCode.clientSessionHasBeenClosed);
    }
    stateStore.tcp!.write(_heartbeatRequest);
    final completer = Completer<void>();
    _heartbeatCompleters.add(completer);
    return completer.future;
  }

  void resolveHeartbeatCompleters() {
    _heartbeatCompleters.removeWhere((completer) {
      completer.complete();
      return true;
    });
  }

  bool rejectHeartbeatCompletersIfFail(TurmsNotification notification) {
    if (_heartbeatFailureRequestId == notification.requestId.toInt()) {
      _rejectHeartbeatCompleters(
          ResponseException.fromNotification(notification));
      return true;
    }
    return false;
  }

  void _rejectHeartbeatCompleters(ResponseException exception) {
    _heartbeatCompleters.removeWhere((completer) {
      completer.completeError(exception);
      return true;
    });
  }

  @override
  Future<void> close() {
    onDisconnected();
    return Future.value();
  }

  @override
  void onDisconnected({Object? error, StackTrace? stackTrace}) {
    stop();
    final exception = ResponseException(
        code: ResponseStatusCode.clientSessionHasBeenClosed,
        cause: error,
        stackTrace: stackTrace);
    _rejectHeartbeatCompleters(exception);
  }
}
